mitiq/collector.py

```diff
@@ -0,0 +1,263 @@
+            circuits and runs a minimal set.
+        """
+        if force_run_all:
+            to_run = circuits
+        else:
+            collection = CircuitCollection(circuits)
+            to_run = collection.unique
+
+        if not self._can_batch:
+            for circuit in to_run:
+                self._call_executor(circuit)
+
+        else:
+            if self._max_batch_size >= len(to_run):
+                self._call_executor(to_run)
+
+            else:
+                stop = len(to_run)
+                step = self._max_batch_size
+                for i in range(int(np.ceil(stop / step))):
+                    batch = to_run[i * step: (i + 1) * step]
+                    self._call_executor(batch)
+
+        # Expand computed results to all results using counts.
+        if force_run_all:
+            return self._computed_results
+        results = []
+        for value, mult in zip(self._computed_results, collection.counts()):
+            results += [value] * mult
+        return results
+
+    def _call_executor(
+        self, to_run: Union[QPROGRAM, Sequence[QPROGRAM]]
+    ) -> None:
+        """Calls the executor on the input circuit(s) to run.
+
+        Args:
+            to_run: Circuit(s) to run.
+        """
+        result = self._executor(to_run)
+        self._calls_to_executor += 1
+
+        try:
+            result = list(result)
+            self._computed_results += result
+            self._executed_circuits += to_run
+        except TypeError:
+            self._computed_results.append(result)
+            self._executed_circuits.append(to_run)
+
+    @staticmethod
+    def is_batched_executor(executor: Callable) -> bool:
+        """Returns True if the input function is recognized as a "batched
+        executor", else False.
+
+        The executor is detected as "batched" if and only if it is annotated
+        with a return type that is one of the following:
+
+            * Iterable[float]
+            * List[float]
+            * Sequence[float]
+            * Tuple[float]
+            * numpy.ndarray
+
+        Batched executors can run several quantum programs in a single call.
+        See below.
+
+        Args:
+            executor: A "serial executor" (1) or a "batched executor" (2).
+
+                (1) A function which inputs a single `QPROGRAM` and outputs a
+                single expectation value as a float.
+                (2) A function which inputs a list of `QPROGRAM`s and outputs a
+                list of expectation values (one for each `QPROGRAM`).
+
+        Returns:
+            True if the executor is detected as batched, else False.
+        """
+        executor_annotation = inspect.getfullargspec(executor).annotations
+
+        return executor_annotation.get("return") in (
+            List[float],
+            Sequence[float],
+            Tuple[float],
+            Iterable[float],
+            np.ndarray,
+        )
+
+
+class CircuitCollection:
```

Comment:
This will be more of a comment rather than a suggestion, but wanted to follow up on what I mentioned on the call yesterday @rmlarose @andreamari 

I had been thinking that using a built-in Python `Counter` object i.e. https://pymotw.com/2/collections/counter.html could simplify this. However, it seems that `Counter` objects don't take a custom equality method and instead rely on the `__hash__` function on the relevant objects ([example](https://stackoverflow.com/questions/14667332/object-equallity-in-python-collections-counter)).

Unfortunately, cirq circuits don't currently hash. Since operations like this become a lot simpler if they do, perhaps we should propose upstreaming a `__hash__` method to cirq that hashes the DAG of the `Circuit`. What do you two think? I would open a separate issue for this and it wouldn't block this issue.